components/server/src/api/server.ts

```diff
@@ -208,39 +208,43 @@ export class API {
                         }
                     };
 
-                    const apply = async <T>(): Promise<T> => {
-                        const subjectId = await self.verify(context);
-                        await rateLimit(subjectId);
-                        context.user = await self.ensureFgaMigration(subjectId);
+                    return withRequestContext(async () => {
+                        const userId = await self.verify(connectContext);
+                        await rateLimit(userId);
+                        await self.ensureFgaMigration(userId);
 
-                        return Reflect.apply(target[prop as any], target, args);
-                    };
-                    if (grpc_type === "unary" || grpc_type === "client_stream") {
-                        return withRequestContext(async () => {
-                            try {
-                                const promise = await apply<Promise<any>>();
-                                const result = await promise;
-                                done();
-                                return result;
-                            } catch (e) {
-                                handleError(e);
-                            }
-                        });
-                    }
-                    return wrapAsyncGenerator(
-                        (async function* () {
-                            try {
-                                const generator = await apply<AsyncGenerator<any>>();
```

Comment:
@akosyakov This is the biggest change here: before we would call `verify` (authentication + ratelimit) on every generation. I'm pretty sure we don't want authentication in there. But maybe ratelimiting? WDYT? :shrug:

Reply:
it is addressed now, right? Or misunderstand something.

I think we call auth only once before creating a generator

Reply:
Pushed a fix.